crates/matrix-sdk-common/src/linked_chunk/builder.rs

```diff
@@ -274,10 +274,179 @@ impl<const CAP: usize, Item, Gap> LinkedChunkBuilder<CAP, Item, Gap> {
+        {
+            // The number of items is not too large.
+            if let ChunkContent::Items(items) = &chunk.content {
+                if items.len() > CAP {
+                    return Err(LinkedChunkBuilderError::ChunkTooLarge { id: chunk.identifier });
+                }
+            }
+
+            // Chunk has no next chunk.
+            if chunk.next.is_some() {
+                return Err(LinkedChunkBuilderError::ChunkIsNotLast { id: chunk.identifier });
+            }
+        }
+
+        // Create the `LinkedChunk` from a single chunk.
+        {
+            // This is the only chunk. Pretend it has no previous chunk if any.
+            chunk.previous = None;
+
+            // Transform the `RawChunk` into a `Chunk`.
+            let chunk_ptr = Chunk::new_leaked(chunk.identifier, chunk.content);
+
+            Ok(Some(LinkedChunk {
+                links: Ends { first: chunk_ptr, last: None },
+                chunk_identifier_generator,
+                updates: Some(ObservableUpdates::new()),
+                marker: PhantomData,
+            }))
+        }
+    }
+
+    pub fn insert_new_first_chunk(
+        linked_chunk: &mut LinkedChunk<CAP, Item, Gap>,
+        new_first_chunk: RawChunk<Item, Gap>,
+    ) -> Result<&Chunk<CAP, Item, Gap>, LinkedChunkBuilderError>
+    where
+        Item: Clone,
+        Gap: Clone,
+    {
+        // Check `LinkedChunk` is going to be consistent after the insertion.
+        {
+            // The number of items is not too large.
+            if let ChunkContent::Items(items) = &new_first_chunk.content {
+                if items.len() > CAP {
+                    return Err(LinkedChunkBuilderError::ChunkTooLarge {
+                        id: new_first_chunk.identifier,
+                    });
+                }
+            }
+
+            // New chunk has no previous chunk.
+            if new_first_chunk.previous.is_some() {
+                return Err(LinkedChunkBuilderError::ChunkIsNotFirst {
+                    id: new_first_chunk.identifier,
+                });
+            }
+
+            let expected_next_chunk = linked_chunk.links.first_chunk().identifier();
+
+            // New chunk has a next chunk.
+            let Some(next_chunk) = new_first_chunk.next else {
+                return Err(LinkedChunkBuilderError::MissingNextChunk {
+                    id: new_first_chunk.identifier,
+                });
+            };
+
+            // New chunk has a next chunk, and it is the first chunk of the `LinkedChunk`.
+            if next_chunk != expected_next_chunk {
+                return Err(LinkedChunkBuilderError::CannotConnectTwoChunks {
+                    new_chunk: next_chunk,
+                    with_chunk: expected_next_chunk,
+                });
+            }
+
+            // Alright. It's not possible to have a cycle within the chunks or
+            // multiple connected components here. All checks are made.
```

Comment:
It's possible to have a cycle in multiple steps, if you start with a linked chunk with ChunkIdentifier (cid) 0, then `insert_new_first_chunk(, { prev: cid(1) })`, then `insert_new_first_chunk(, { prev: cid(0) })`?